File: 01_introduction_to_algorithms/01_binary_search.scala
def binarySearch(sortedList: List[Int], item: Int): Int = {
  var low = 0
  var high = sortedList.length
  println(low, high)
  low
}

val myList = List(1,3,5,7,9)
println(binarySearch(myList, 3))
println(binarySearch(myList, -1))

File: 01_introduction_to_algorithms/01_binary_search.ts
function binarySearch<T>(sortedList: T[], item: T): number | null {
  let left = 0
  let right = sortedList.length - 1

  while (left <= right) {
    const mid = Math.floor((left + right) / 2)
    const guess: T = sortedList[mid]

    if (guess === item) return mid
    if (guess > item) {
      right = mid - 1
    } else {
      left = mid + 1
    }
  }

  return null
}

const list1 = [1, 3, 5, 7, 9, 11, 13, 15]

console.log(binarySearch(list1, 3)) // 1
console.log(binarySearch(list1, 9)) // 4
console.log(binarySearch(list1, -1)) // null
